src/scene/light.js

```diff
@@ -698,7 +714,17 @@ class Light {
         const g = color.g;
         const b = color.b;
 
-        const i = this._intensity;
+        let i = this._intensity;
+        if (this._scene?.physicalUnits) {
+            if (this._type === LIGHTTYPE_SPOT) {
+                const angleAsRadians = this._outerConeAngle * Math.PI / 180.0;
+                i = this._luminance / (2 * Math.PI * (1 - Math.cos(angleAsRadians / 2.0)));
```

Comment:
I'm not sure but I think we have to use "_innerConeAngle" here.

Maybe `i = this._luminance / (2 * Math.PI * (1 - (this._innerConeAngleCos + this._outerConeAngleCos)  / 2.0)));`. 
Probably, it is wrong.

And I'm just curious why do we use `Math.cos(angleAsRadians / 2.0)` and not `Math.cos(angleAsRadians) / 2.0`

Reply:
Ah, so the outer cone angle controls how much energy should transmit through the light volume, such that we can convert the light in lm to lm/m^2, so it would only make sense to use the outer angle as the inner angle is only used to feather the edge of the light. 

As for the math, I would defer to the [filament documentation](https://google.github.io/filament/Filament.md.html#lighting/units/lightunitsvalidation:~:text=light%27s%20direction%20vector-,Equation,-60).

Reply:
It sounds a little strange to me. Energy cannot be equal for **innerConeAngle === 0** and **innerConeAngle === outerConeAngle**. Filament evaluates this as if the spot light has uniformly distributed intensity.
<img width="456" alt="Screen Shot 2022-09-07 at 00 20 24" src="https://user-images.githubusercontent.com/104348270/188730834-495e91e3-91f2-445d-8552-de7db6c1847f.png">

It may be true from 0 to **innerConeAngle**. But  between **innerConeAngle** and **outerConeAngle** we have the angular falloff attenuation (`(cos(angle) - _outerConeAngleCos) / (_innerConeAngle - _outerConeAngleCos)` or something like this). So, it looks like multiplier has to use linear combination of **cos(innerConeAngle)** and **cos(outerConeAngle)**.

Reply:
This is what part uses: https://github.com/mmp/pbrt-v4/blob/faac34d1a0ebd24928828fe9fa65b65f7efc5937/src/pbrt/lights.cpp#L1463

Reply:
This is what part uses: https://github.com/mmp/pbrt-v4/blob/faac34d1a0ebd24928828fe9fa65b65f7efc5937/src/pbrt/lights.cpp#L1463

Reply:
When thinking about this a bit, the inner angle basically defines a border outside of which the spot light cone is in the shape of solid angle, where no inner == outer means we have a perfect cone and inner == 0 means we have a perfect solid angle. If we take that into account, the amount of energy in the volume is dependent on the inner angle, where there is a bit of energy lost if the inner angle is less than the outer. 

In these illustrations I use a hemisphere to indicate the solid angle cutout of the spot light. I hope they make sense.

Solid angle when outer == inner
![image](https://user-images.githubusercontent.com/107400752/189383966-4a7ccde4-fccc-4006-bf65-d11ae8a35250.png)

Solid angle when inner == 0.
![image](https://user-images.githubusercontent.com/107400752/189388034-060fe372-bd4c-4530-bbb9-a1386650006d.png)

Solid angle when inner > 0 && inner < outer
![image](https://user-images.githubusercontent.com/107400752/189388132-85e8a69d-a6ee-4ec2-96f7-5a4e08a5266e.png)

So we can definitely look into accounting for that energy loss, I don't think it makes a ton of difference but it's definitely worth looking into.

Reply:
If we assume that the angular falloff attenuation is defined as `(cos(angle) - _outerConeAngleCos) / (_innerConeAngle - _outerConeAngleCos)`, then the complete integral should look like [this](https://www.wolframalpha.com/input?i2d=true&i=+phi+%3D+2*pi*%5C%2840%29Integrate%5BI%2C%7Bx%2Ccos%5C%2840%29a%5C%2841%29%2C1%7D%5D+%2BIntegrate%5BDivide%5B%5C%2840%29cos%5C%2840%29x%5C%2841%29+-+cos%5C%2840%29a%5C%2841%29%5C%2841%29%2C%5C%2840%29cos%5C%2840%29b%5C%2841%29+-+cos%5C%2840%29a%5C%2841%29%5C%2841%29%5D*+I%2C%7Bx%2Ccos%5C%2840%29b%5C%2841%29%2Ccos%5C%2840%29a%5C%2841%29%7D%5D%5C%2841%29)

Where
* **_a_** is `this._innerConeAngle * Math.PI / 360.0`
* **_b_** is `this._outerConeAngle * Math.PI / 360.0`,
* **the first integral** is about everything that inside the inner cone,
* **the second integral** is about everything that between the inner cone and the outer cone

<img width="509" alt="Screen Shot 2022-09-10 at 03 20 25" src="https://user-images.githubusercontent.com/104348270/189457940-e4290af6-dfa2-44f1-a7ce-b802ab837e9b.png">

And, probably, the final result is something like this

```js
i = this._luminance / (
    2 * Math.PI * I * (
        1 + (
            Math.sin(Math.cos(b)) - Math.sin(Math.cos(a))
        ) / (
            Math.cos(a) - Math.cos(b)
        )
    )
);
```

But, I may miss something, so it is better to check it.

Reply:
I constructed two plots for two angular falloff attenuation models.

1) [Cosine attenuation](https://www.desmos.com/calculator/tw4hncxrx0?lang=ru) + [smoothstep](https://github.com/playcanvas/engine/blob/6424d6786c5f78853221b7177dc9da887fda410b/src/graphics/program-lib/chunks/lit/frag/spot.js#L4): `(cos(angle) - _outerConeAngleCos) / (_innerConeAngleCos - _outerConeAngleCos)`
<img width="572" alt="Screen Shot 2022-09-10 at 15 11 37" src="https://user-images.githubusercontent.com/104348270/189480647-b0b18df2-b79f-4187-88b2-064c349068ac.png">

2) [Linear attenuation](https://www.desmos.com/calculator/tzuisqed11?lang=ru)  + [smoothstep](https://github.com/playcanvas/engine/blob/6424d6786c5f78853221b7177dc9da887fda410b/src/graphics/program-lib/chunks/lit/frag/spot.js#L4): `(angle- _outerConeAngle) / (_innerConeAngle - _outerConeAngle)`
<img width="572" alt="Screen Shot 2022-09-10 at 15 11 33" src="https://user-images.githubusercontent.com/104348270/189480652-6ed37dca-5ee3-4c1a-a147-b1b31dcbabb4.png">

Dash lines show luminous intensity factor. As you can see on the second model, if we shift the left side on the right and sum area with "attenuations", then we get constant 1. And this means that combinations of two "attenuations" equal area inside innerConeAngle. So for linear attenuation we get exactly what is used in PBR book:

```js
const innerAngleAsRadians = this._innerConeAngle * Math.PI / 180.0;
const outerAngleAsRadians = this._outerConeAngle * Math.PI / 180.0;
i = this._luminance / (
    2 * Math.PI * (
        1 - Math.cos(innerAngleAsRadians) + (Math.cos(innerAngleAsRadians) -  Math.cos(outerAngleAsRadians)) / 2
    )
);
```

Reply:
@GSterbrant
I don't know if it is correct. So, [here is an issue in FIlament](https://github.com/google/filament/issues/6060). Since, they don't use the **[smoothstep](https://github.com/playcanvas/engine/blob/6424d6786c5f78853221b7177dc9da887fda410b/src/graphics/program-lib/chunks/lit/frag/spot.js#L4)** function their falloff attenuation function is linear.